// @inp-net/churros-client/build/index.js
export class ChurrosClient {
    clientId;
    clientSecret;
    redirectURI;
    state;
    token;
    constructor({ client_id, client_secret, redirect_uri, }) {
        this.clientId = client_id;
        this.clientSecret = client_secret;
        this.redirectURI = redirect_uri;
    }
    generateState() {
        this.state = Math.random().toString(36).slice(2, 15);
    }
    get redirectURL() {
        return new URL(this.redirectURI);
    }
    get authorizationURL() {
        if (!this.state)
            this.generateState();
        return `https://churros.inpt.fr/authorize?${new URLSearchParams({
            client_id: this.clientId,
            redirect_uri: this.redirectURI,
            response_type: 'code',
            state: this.state ?? '',
        }).toString()}`;
    }
    get tokenURL() {
        return `https://churros.inpt.fr/token`;
    }
    get userInfoURL() {
        return `https://churros.inpt.fr/identity`;
    }
    async getToken(code, state) {
        if (state !== this.state)
            throw new Error('Invalid state, potential CSRF attack');
        const response = await fetch(this.tokenURL, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded',
                'Authorization': 'Basic ' + btoa(this.clientId + ':' + this.clientSecret),
            },
            body: new URLSearchParams({
                grant_type: 'authorization_code',
                code: code,
                redirect_uri: this.redirectURI,
            }),
        });
        const { access_token, token_type } = (await response.json());
        if (token_type !== 'bearer')
            throw new Error('Invalid token type, expected bearer');
        return (this.token = access_token);
    }
    async getUserInfo(token) {
        const response = await fetch(this.userInfoURL, {
            method: 'GET',
            headers: {
                Authorization: 'Bearer ' + token,
            },
        });
        return (await response.json());
    }
}
